Lodestar prover for execution api

nazarhussain opened the pull request:
**Motivation**

Make it easier for third party developers to utilize the light client and verify the non-trusted eth responses. 

**Description**

Create a lightclient prover utility for web3 providers. 

Closes #4706

**Steps to test or reproduce**

Run all tests. 

```sh
git checkout <feature_branch>
lodestar beacon --new-flag option1
```

nazarhussain committed:
Add package skeleton

nazarhussain committed:
Add implementation for the verification

nazarhussain committed:
Update package versions

dapplion reviewed (commented):
Overall direction looks good! :fire: looking forward to the polished PR

nazarhussain committed:
Update teh provdier structure to store full execution payload

nazarhussain committed:
Add a test script

nazarhussain committed:
Split the utils to scoped files

nazarhussain committed:
Add multiple web3 providers

nazarhussain commented:
This PR functionality is ready and tested manually with the example given in the readme file. Now adding the unit and integration tests. 

Please share your earlier feedback.

nazarhussain committed:
Add a proxy factory method

nazarhussain committed:
Add the CLI for the prover proxy

nazarhussain committed:
Rename few functions to make those consistent

nazarhussain committed:
Add some required unit tests

nazarhussain committed:
Add unit tests

nazarhussain committed:
Add e2e tests

nazarhussain committed:
Fix duplicate Buffer error

nazarhussain committed:
Fix lint error

nazarhussain committed:
Fix lint errors

nazarhussain committed:
Update the lightclient to sync in background

nazarhussain committed:
Validate the execution payload

g11tech commented on packages/prover/src/proof_provider/proof_provider.ts:
```diff
@@ -0,0 +1,200 @@
+  // Map to match CL slot to EL block number
+  private slotsMap: {[slot: number]: string} = {};
+
+  lightClient?: Lightclient;
```
wouldn't make it optional but assign in constructor

nazarhussain replied:
Can't do that because of the comment below. https://github.com/ChainSafe/lodestar/pull/5222#discussion_r1134191019

g11tech commented on packages/prover/src/proof_provider/proof_provider.ts:
```diff
@@ -0,0 +1,200 @@
+  static buildWithRestApi(urls: string[], opts: RootProviderInitOptions): ProofProvider {
+    const config = createChainForkConfig(networksChainConfig[opts.network]);
+    const api = getClient({urls}, {config});
+    const transport = new LightClientRestTransport(api);
```
should construct the lightclient here itself and send in the constructor

nazarhussain replied:
To construct lightclient we need to call some `async` behavior. I strictly don't want to make `init` as `async` to make it useful in the web3_provider context.

g11tech replied:
you may use .then(lightclient => lightclient.start())

nazarhussain replied:
I don't want to inject the lightclient to proof provider. 

1. Lightclient initialization is strictly an `async` process because of some BLS issue documented there. So don't want to use it in a strictly non-async `init` method of proof provider. 
2. Directly calling the `Lightclient` constructor would require to duplicate some logic used in light client initialization. 

I think we should proceed with the current implementation as of now. Once we add other transport modes, there may appear a need refactor that logic.

g11tech commented on packages/prover/src/proof_provider/proof_provider.ts:
```diff
@@ -0,0 +1,200 @@
+    return this.readyPromise;
+  }
+
+  static buildWithRestApi(urls: string[], opts: RootProviderInitOptions): ProofProvider {
```
change it into static method `init` and take the transport type as an option and do if then else for transport below and throw not implemented error for p2p

g11tech commented on packages/prover/src/proof_provider/proof_provider.ts:
```diff
@@ -0,0 +1,200 @@
+      config,
+      api,
+      transport,
+    });
```
call sync from here itself than in the constructor

g11tech replied:
call sync from here itself than in the constructor and assign to the ready promise

g11tech commented on packages/prover/src/proof_provider/proof_provider.ts:
```diff
@@ -0,0 +1,200 @@
+    this.registerEvents();
+
+    const headSlot = this.lightClient.getHead().beacon.slot;
+    await this.lightClient?.sync(
```
isn't what lightClient.start does?

nazarhussain replied:
No. I believe `start` and `sync` are doing two different things.

g11tech replied:
start does sync and much more, so external sync is not required.
you need to return a promise that resolves when lightClient.status is shifted to started, for that you may need to emit a syncronized event in the lightclient which will resolve this promise

g11tech commented on packages/prover/src/proof_provider/proof_provider.ts:
```diff
@@ -0,0 +1,200 @@
+    const checkpointRoot = await getSyncCheckpoint(api, checkpoint);
+    const genesisData = await getGenesisData(api);
+
+    this.lightClient = await Lightclient.initializeFromCheckpointRoot({
```
this should just happen in init and just start should be called

readyPromise is a wait on the lightClient.status to turn into `started`

nazarhussain replied:
I want the `ProofProvider` ready to when it's synced.

g11tech commented on packages/prover/src/proof_provider/proof_provider.ts:
```diff
@@ -0,0 +1,200 @@
+
+    // Load the payloads from the CL
+    const startSlot = headSlot;
+    const endSlot = headSlot - MAX_PAYLOAD_HISTORY;
```
you can't go beyond the finalized checkpoint's epoch as the beacon node itself might not have that data, so a max condition also has to be applied unless max payload history < 2 epochs

g11tech commented on packages/prover/src/proof_provider/proof_provider.ts:
```diff
@@ -0,0 +1,200 @@
+    // Load the payloads from the CL
+    const startSlot = headSlot;
+    const endSlot = headSlot - MAX_PAYLOAD_HISTORY;
+    const payloads = await getExecutionPayloads(this.options.api, startSlot, endSlot);
```
would it be beneficial to fetch them dynamically on requirement? also why slot => endslot mapping is required?

may be we just need to track head, and finalized slot's number?

g11tech commented on packages/prover/src/proof_provider/proof_provider.ts:
```diff
@@ -0,0 +1,200 @@
+    const finalizedSlot = this.lightClient.getFinalized().beacon.slot;
+    const finalizedPayload = await getExecutionPayloads(this.options.api, finalizedSlot, finalizedSlot);
+    const finalizedBlockNumber = numberToHex(finalizedPayload[finalizedSlot].blockNumber);
+    this.finalizedPayloads.set(finalizedBlockNumber, finalizedPayload[finalizedSlot]);
```
again why do we need this mapping, once a slot is finalized, all numbers corresponding to finalized's block number and before are finalized, why do we need this mapping

nazarhussain committed:
Update initWithRest to init

g11tech commented on packages/prover/src/proof_provider/proof_provider.ts:
```diff
@@ -0,0 +1,200 @@
+    throw new Error(`Invalid blockNumber "${blockNumber}"`);
+  }
+
+  async update(update: allForks.LightClientHeader, finalized = false): Promise<void> {
```
lets rename the arg `update` here to  or `lcHeader`

g11tech commented on packages/prover/src/proof_provider/proof_provider.ts:
```diff
@@ -0,0 +1,200 @@
+      return;
+    }
+
+    if (isForkExecution(fork) && !("execution" in update)) {
```
another check you need to make is that execution is not the defaultValue() as the upgraded `lcHeader` will have `defaultValue()` assigned to execution

nazarhussain replied:
Do we need to be worried if `execution` have default value as long as it's an execution fork?

g11tech replied:
yes, because execution is actually being bundled on/post `capella` only in the updates

nazarhussain committed:
Update the max limit for the payloads to not cross finalized slot

g11tech commented on packages/prover/src/proof_provider/proof_provider.ts:
```diff
@@ -0,0 +1,200 @@
+    });
+
+    this.lightClient.emitter.on(LightclientEvent.lightClientFinalityUpdate, async (data) => {
+      await this.update(data, true).catch((e) => {
```
aren't you `typing` arg of lightclient header to this.update, data you should get here is of type update

Also it would be better to emit `latest` and `finalized` events from lightclient as thats what you mean to process here i think

nazarhussain replied:
I didn't get that comment. Please explain it a bit.

g11tech replied:
https://github.com/ChainSafe/lodestar/pull/5222/files#diff-229bc1a2e4f16b99b96ae3bd25f0429ff996cf967786ac6235f53f8f4a5b1e2cR142 
not sure the types match

nazarhussain replied:
Yes the type matches to the `lightclient` events else TS would not have let pass it. Remember it's not the direct event handler, rather a function called from within event handler without any type casting.

g11tech replied:
ahh, so this event comes from light-client package  which has same name as one in beacon-node

I would tend to rename these events in light-client package as: `lightClientOptimisticHeader` and `lightClientFinalityHeader` and rename `this.update` to `this.processUpdateHeader`

nazarhussain committed:
Remove the usage for finalizedPayloadHeaders tracking

nazarhussain committed:
Rename update to lcHeader

g11tech commented on packages/prover/src/utils/execution.ts:
```diff
@@ -0,0 +1,108 @@
+    storageRoot: proof.storageHash,
+    codeHash: proof.codeHash,
+  });
+  return account.serialize().equals(expectedAccountRLP ? expectedAccountRLP : emptyAccountSerialize);
```
this looks redundant, verify proof already checks for match of these things i think, will double check and revert what the correct condition should be

nazarhussain replied:
I believe this is additional check for `storageHash` and `codeHash` because `trie.verifyProof` does not check both. I took this logic from `patronum` referred that references in our code.

g11tech commented:
looks exciting on the first pass :clap: , will need to give more brain/review cycles to the PR

g11tech commented on packages/prover/src/cli/cmds/start/index.ts:
```diff
@@ -0,0 +1,17 @@
+import {startHandler} from "./handler.js";
+import {StartArgs, startOptions} from "./options.js";
+
+export const start: CliCommand<StartArgs> = {
```
can we rename it to `proverProxyHandler` or something more meaningful

g11tech commented on packages/prover/src/interfaces.ts:
```diff
@@ -0,0 +1,51 @@
+import {ProofProvider} from "./proof_provider/proof_provider.js";
+import {ELRequestPayload, ELResponse} from "./types.js";
+
+export enum LightNode {
```
LCTransport is more appropriate name

g11tech commented on packages/prover/src/cli/cmds/start/options.ts:
```diff
@@ -0,0 +1,81 @@
+  port: number;
+  network: string;
+  "execution-rpc": string;
+  mode: "rest" | "p2p";
```
would like to use transport as the arg name

g11tech commented on packages/prover/src/cli/cmds/start/options.ts:
```diff
@@ -0,0 +1,81 @@
+export type StartArgs = {
+  port: number;
+  network: string;
+  "execution-rpc": string;
```
use use `executionUrls` and make it an array which we may support later

nazarhussain replied:
Renamed to `execution-url`. I don't see a reason to have it array for now. If we make it plural for now user can mislead to behavior.

g11tech commented on packages/prover/src/cli/cmds/start/options.ts:
```diff
@@ -0,0 +1,81 @@
+  network: string;
+  "execution-rpc": string;
+  mode: "rest" | "p2p";
+  "beacon-rpc"?: string[];
```
would prefer beaconUrls

g11tech commented on packages/prover/src/cli/cmds/start/options.ts:
```diff
@@ -0,0 +1,81 @@
+
+export type StartOptions = {
+  network: NetworkName;
+  executionRpcUrl: string;
```
keep them same name as in args for easy correlation

g11tech commented on packages/prover/src/cli/cmds/start/options.ts:
```diff
@@ -0,0 +1,81 @@
+  mode: LightNode;
+  beaconRpcUrls: string[];
+  beaconBootNodes: string[];
+  checkpoint?: string;
```
would prefer wsCheckpoint to reflect it shouldn't be older than weak subjectivity period, also reflect the same in description

g11tech commented on packages/prover/tsconfig.build.json:
```diff
@@ -0,0 +1,9 @@
+  "include": ["src"],
+  "compilerOptions": {
+    "outDir": "lib",
+    // Had to add it because compiling error `Duplicate identifier 'Buffer'`
```
i think the commend here will break json or just check why github is showing it red

nazarhussain replied:
Github parse JSON as normal json files where comments are not allowed. And TS parse these as JSON5 files where comments are allowed.

g11tech commented on packages/prover/src/proof_provider/proof_provider.ts:
```diff
@@ -0,0 +1,205 @@
+    throw new Error(`Invalid blockNumber "${blockNumber}"`);
+  }
+
+  async update(lcHeader: allForks.LightClientHeader, finalized = false): Promise<void> {
```
rename to `processUpdateHeader` for better clarity or `processLCHeader`

g11tech commented on packages/prover/src/proof_provider/proof_provider.ts:
```diff
@@ -0,0 +1,205 @@
+  async update(lcHeader: allForks.LightClientHeader, finalized = false): Promise<void> {
+    const fork = this.options.config.getForkName(lcHeader.beacon.slot);
+
+    if (!isForkExecution(fork)) {
+      return;
+    }
+
+    if (isForkExecution(fork) && !("execution" in lcHeader)) {
+      throw new Error("Execution payload is required for execution fork");
+    }
```
!isForkExecution has already been checked

nazarhussain replied:
No it's not checked already.

g11tech commented on packages/prover/src/proof_provider/proof_provider.ts:
```diff
@@ -0,0 +1,205 @@
+    for (const [slot, payload] of Object.entries(payloads)) {
+      const blockNumber = numberToHex(payload.blockNumber);
+      this.payloads.set(blockNumber, payload);
+      this.slotsMap[parseInt(slot)] = blockNumber;
```
i don't understand slot to number mapping, all numbers <= finalized's number are finalized, all numbers <= head's number are present

if there is a mapping that needs to be stored is beacon block root => execution block root for the non-finalized set and tracking what is head block

g11tech commented on packages/prover/src/proof_provider/proof_provider.ts:
```diff
@@ -0,0 +1,205 @@
+    const finalizedPayload = await getExecutionPayloads(this.options.api, finalizedSlot, finalizedSlot);
+    const finalizedBlockNumber = numberToHex(finalizedPayload[finalizedSlot].blockNumber);
+    this.finalizedSlot = finalizedSlot;
+    this.slotsMap[finalizedSlot] = finalizedBlockNumber;
```
track finalized slot => number mapping separately and thats the only slot => number reference that makes sense

g11tech commented on packages/prover/src/proof_provider/proof_provider.ts:
```diff
@@ -0,0 +1,205 @@
+  }
+
+  getStatus(): {latest: number; finalized: number; status: RunStatusCode} {
+    assertLightClient(this.lightClient);
```
instead of throwing on unitialized light client, have an uninitialized status

g11tech commented on packages/prover/src/proof_provider/proof_provider.ts:
```diff
@@ -0,0 +1,205 @@
+      throw new Error("Execution payload is required for execution fork");
+    }
+
+    const newPayload = (lcHeader as capella.LightClientHeader).execution;
```
rename it to newPayloadHeader for clarity as this is not a full payload with transactions

g11tech commented on packages/prover/src/proof_provider/proof_provider.ts:
```diff
@@ -0,0 +1,205 @@
+    const blockSlot = lcHeader.beacon.slot;
+    const existingPayload = this.payloads.get(blockNumber);
+
+    if (existingPayload && existingPayload.blockHash === newPayload.blockHash) {
```
change it into nested if with outer if (existingPayload!==undeined)
and inner an if else

g11tech commented on packages/prover/src/proof_provider/proof_provider.ts:
```diff
@@ -0,0 +1,205 @@
+    const newPayload = (lcHeader as capella.LightClientHeader).execution;
+    const blockNumber = numberToHex(newPayload.blockNumber);
+    const blockSlot = lcHeader.beacon.slot;
+    const existingPayload = this.payloads.get(blockNumber);
```
for non finalized, only maintain mappings of roots to roots and refrain from using mapping of numbers as they are susceptible to reorg

g11tech commented on packages/prover/src/proof_provider/proof_provider.ts:
```diff
@@ -0,0 +1,205 @@
+      return;
+    }
+
+    if (existingPayload && existingPayload.blockHash !== newPayload.blockHash) {
```
if you refer unfinalized by hashes and not numbers this problem will not arise

g11tech commented on packages/prover/src/proof_provider/proof_provider.ts:
```diff
@@ -0,0 +1,205 @@
+      // 3. or update the blockHash
+    }
+
+    const newPayloads = await getExecutionPayloads(this.options.api, blockSlot, blockSlot);
```
newPayloadHeaders

g11tech commented on packages/prover/src/utils/consensus.ts:
```diff
@@ -0,0 +1,87 @@
+  throw new Error(`Can not fetch nearest block for slot=${slot}`);
+}
+
+export async function getSlotLimitsForPayloads(lightClient: Lightclient): Promise<{start: number; end: number}> {
```
could be renamed to getUnFinaliziedRange

g11tech commented on packages/prover/src/utils/consensus.ts:
```diff
@@ -0,0 +1,87 @@
+  payloads[block.message.slot] = block.message.body.executionPayload;
+
+  while (slot >= startSlot) {
+    const previousBlock = (await fetchNearestBlock(api, block.message.slot - 1, "down")) as capella.SignedBeaconBlock;
```
you should do an array fetch here may be on a batch of request promises (for loop for batching in batch size for e.g.)

nazarhussain replied:
I had added an extra task to epic to refactor this logic to use the batch RPC requests.

g11tech commented on packages/prover/src/utils/consensus.ts:
```diff
@@ -0,0 +1,87 @@
+    const previousBlock = (await fetchNearestBlock(api, block.message.slot - 1, "down")) as capella.SignedBeaconBlock;
+
+    if (block.message.body.executionPayload.parentHash === previousBlock.message.body.executionPayload.blockHash) {
+      payloads[block.message.slot] = block.message.body.executionPayload;
```
this will easily break on reorgs, just store unfinalized payloads referenced by hash

only condition you need to check in the end if if finalized is ancesor of the head

nazarhussain replied:
that's just an extra safety check while fetching the blocks from the API. It can be skipped. The re-org scenario is covered on different place.

g11tech replied:
yes just skip the check as anyway this mapping is not canonical since it unfinalized and will cause intractable logical errors

g11tech commented on packages/prover/test/e2e/web3_provider.test.ts:
```diff
@@ -0,0 +1,45 @@
+    describe("ethers", () => {
+      it("should connect to the network and call non-verified method", async () => {
+        const {provider} = createVerifiedExecutionProvider(
+          new ethers.JsonRpcProvider("https://lodestar-sepoliarpc.chainsafe.io"),
```
i am not sure if we should rely on connecting to real world servers as the server uptimes can't be guaranteed espeicially during client software upgrades

nazarhussain replied:
I had added it as task in the main issue to setup a local EL/CL servers for CI.

g11tech commented on packages/prover/src/web3_proxy.ts:
```diff
@@ -0,0 +1,122 @@
+  }
+
+  const proxyServer = http.createServer(function proxyRequestHandler(req, res) {
+    if (req.url === "/proxy") {
```
would like to keep the url as `/prover` for better presentation

nazarhussain replied:
The `/` is the actual prover url and `/proxy` is just forwarding endpoint. So this seems ok to me.

g11tech commented on packages/prover/src/verified_requests/eth_getBalance.ts:
```diff
@@ -0,0 +1,29 @@
+    params: [address, block],
+  } = payload;
+  const executionPayload = rootProvider.getExecutionPayload(block ?? "latest");
+  const proof = await getELProof(handler, [address, [], numberToHex(executionPayload.blockNumber)]);
```
use hash not number to fetch

nazarhussain committed:
Update the code as per feedback

nazarhussain committed:
Fix readme

nazarhussain committed:
Update the payload store logic

nazarhussain committed:
Add the cleanup logic to payload store

g11tech commented on packages/prover/src/cli/cmds/start/options.ts:
```diff
@@ -0,0 +1,81 @@
+  "ws-checkpoint"?: string;
+};
+
+export type StartOptions = {
```
can we change this to {...} & ({ transport:LCTransport.rest,beaconUrls:..} | {transport:LCTransport.p2p,beaconBootnodes:...}

g11tech commented on packages/prover/src/interfaces.ts:
```diff
@@ -0,0 +1,51 @@
+  signal: AbortSignal;
+  config?: ChainForkConfig;
+  wsCheckpoint?: string;
+} & ({transport: LCTransport.Rest; urls: string[]} | {transport: LCTransport.P2P; bootnodes: string[]});
```
lets keep the same name of `urls` or `bootnodes` in cli/options and here as well so as to just have a clear correspondence that they are all the same

g11tech commented on packages/prover/src/proof_provider/payload_store.ts:
```diff
@@ -0,0 +1,154 @@
+ */
+export class PayloadStore {
+  // We store the block numbers only for finalized blocks
+  private finalizedRoots = new OrderedMap<BlockELRoot>();
```
nice

g11tech commented on packages/prover/src/proof_provider/payload_store.ts:
```diff
@@ -0,0 +1,154 @@
+  private unfinalizedRoots = new Map<BlockCLRoot, BlockELRoot>();
+
+  // Payloads store with BlockELRoot as key
+  private payloads = new Map<BlockELRoot, allForks.ExecutionPayload>();
```
:+1:

g11tech commented on packages/prover/src/proof_provider/payload_store.ts:
```diff
@@ -0,0 +1,154 @@
+
+  get latest(): allForks.ExecutionPayload | undefined {
+    if (this.latestBlockRoot) {
+      return this.payloads.get(this.latestBlockRoot);
```
:+1:

g11tech commented on packages/prover/src/proof_provider/payload_store.ts:
```diff
@@ -0,0 +1,154 @@
+  async getOrFetchFinalizedPayload(blockNumber: number): Promise<allForks.ExecutionPayload | undefined> {
+    if (blockNumber > this.finalizedRoots.max) {
+      throw new Error(
+        `Block number ${blockNumber} is higher than the latest finalized block number. We recommend to use block hash for unfinalized blocks.`
```
:+1:

g11tech commented on packages/prover/src/proof_provider/payload_store.ts:
```diff
@@ -0,0 +1,154 @@
+      );
+    }
+
+    let blockELRoot = this.finalizedRoots.get(blockNumber);
```
a comment here would be nice:
`\\ check if we have payload cached locally else fetch from api`

g11tech commented on packages/prover/src/proof_provider/payload_store.ts:
```diff
@@ -0,0 +1,154 @@
+  set(payload: allForks.ExecutionPayload, finalized: boolean): void {
+    const blockRoot = bufferToHex(payload.blockHash);
+    this.payloads.set(blockRoot, payload);
+    this.latestBlockRoot = blockRoot;
```
if you are fetching payloads in getOrFetchFinalizedPayload and setting there, the latestBlockRoot is getting updated as well from finalized roots, is this correct behavior?

g11tech commented on packages/prover/src/proof_provider/payload_store.ts:
```diff
@@ -0,0 +1,154 @@
+    }
+  }
+
+  async processLCHeader(header: capella.LightClientHeader, finalized = false): Promise<void> {
```
why are we not using allForks.LightClientHeader?

nazarhussain replied:
Because we are intended to use `capella` upward, earlier forks don't have execution payloads.

g11tech replied:
then use Exclude<allforks, altair|bellatrix>

nazarhussain replied:
We can't use the type namespaces with the `utility types`. I believe Caymen was wroking on type refactor where we can get such extended types. We can revisit this logic when we have those types.

g11tech replied:
i think: `Exclude<allforks.LightClientHeader, altair.LightClientHeader | bellatrix.LightClientHeader>` can be done, no?

g11tech commented on packages/prover/src/proof_provider/payload_store.ts:
```diff
@@ -0,0 +1,154 @@
+
+    if (finalized) {
+      if (existingELRoot) {
+        this.unfinalizedRoots.delete(blockCLRoot);
```
we might need to delete all `unfinalizedRoots` which have slot <= finalized

nazarhussain replied:
We don't store the `slot` numbers for the `unfinalizedRoots` so can't delete those with that condition. I added an action item to revisit that logic for future. Once we add extended unit tests this logic will be easy to verify and refactor.

g11tech replied:
but you have payloads stored for them, so do a look and cleanup

g11tech commented on packages/prover/src/proof_provider/payload_store.ts:
```diff
@@ -0,0 +1,154 @@
+      return;
+    }
+
+    if (existingELRoot && existingELRoot !== blockELRoot) {
```
can we chain all above three `if` as `if else if else if...` for better understanding of the flow?

nazarhussain replied:
I prefer guard conditions over `if else` structure. Had simplified it conditions a bit.

g11tech replied:
nothing else needs to be change, just chain them with `else if`

nazarhussain replied:
I prefer the guard conditions over `if else` that makes the code more readable. I think we should just pass on this formatting preference.

g11tech commented on packages/prover/src/proof_provider/payload_store.ts:
```diff
@@ -0,0 +1,154 @@
+    const blockSlot = header.beacon.slot;
+    const blockNumber = header.execution.blockNumber;
+    const blockELRoot = bufferToHex(header.execution.blockHash);
+    const blockCLRoot = bufferToHex(header.beacon.stateRoot);
```
can we add a jsdoc here that blockCLRoot is tracking a beacon block by stateRoot

nazarhussain replied:
Can understand JSDoc for a local variable? Had added doc comments to class members.

g11tech replied:
yes pls add for more clarity to the code reader (or comment) it took me sometime to follow what is goin on with the condition checks below

nazarhussain replied:
I believe the names of the variables `blockCLRoot, `blockELRoot`, `blockNumber` and `blockSlot` are very clear and self explanatory. No need to add comments for these.

g11tech commented on packages/prover/src/proof_provider/payload_store.ts:
```diff
@@ -0,0 +1,154 @@
+    }
+
+    if (existingELRoot && existingELRoot !== blockELRoot) {
+      // Re-org happened, we need to update the payload
```
If i understand correctly, I think el root by CL stateroot can never differ since stateRoot has executionPayloadHeader 's merkel hash included in it, if this happens, this is failure of consensus and is a critical error and you should just reject the header

nazarhussain replied:
My understanding is, the reorg can happen on the CL and EL side. Each fork on CL can point to different fork on EL. So incase we receive a header update which for which we already have a payload in record but have different root that's a sign of such reorg. In that case we consider new version to be right but yet unfinalized.

g11tech commented on packages/prover/src/proof_provider/payload_store.ts:
```diff
@@ -0,0 +1,154 @@
+    }
+
+    this.payloads.set(blockELRoot, (await getExecutionPayloads(this.opts.api, blockSlot, blockSlot))[blockSlot]);
+    this.latestBlockRoot = blockELRoot;
```
if the event is finalized, then latestBlockRoot should not be updated only if slot >= current latestBlockRoot's slot

g11tech commented on packages/prover/src/proof_provider/proof_provider.ts:
```diff
@@ -0,0 +1,182 @@
+export class ProofProvider {
+  private store: PayloadStore;
+
+  private readyPromise?: Promise<void>;
```
add comment: `make sure readyPromise doesn't throw unhandled exceptions`

g11tech commented on packages/prover/src/proof_provider/proof_provider.ts:
```diff
@@ -0,0 +1,182 @@
+      transport,
+    });
+
+    provider.readyPromise = provider.sync(opts.wsCheckpoint);
```
add a .catch to `provider.sync(...).catch(e=>{})` to make sure it doesn't throw unhandled exceptions

g11tech commented on packages/prover/src/proof_provider/proof_provider.ts:
```diff
@@ -0,0 +1,182 @@
+    // Load the finalized payload from the CL
+    const finalizedSlot = this.lightClient.getFinalized().beacon.slot;
+    const finalizedPayload = await getExecutionPayloads(this.opts.api, finalizedSlot, finalizedSlot);
+    this.store.set(finalizedPayload[finalizedSlot], true);
```
you can use a better pattern of Object.values().forEach(payload=>this.store.set(payload,true/false)) here and everywhere else

nazarhussain replied:
`Object.values` is used where it's intended use. On this line we need to set a single payload so didn't used it.

g11tech commented on packages/prover/src/proof_provider/proof_provider.ts:
```diff
@@ -0,0 +1,182 @@
+      isForkExecution(fork) &&
+      (!("execution" in lcHeader) || lcHeader.execution === ssz.capella.ExecutionPayloadHeader.defaultValue())
+    ) {
+      throw new Error("Execution payload is required for execution fork");
```
you should check for isForkWithdrawals or something here as only they will have a non zero execution, also you can't === objects you will need to use : `ssz.capella.ExecutionPayloadHeader.equals(lcHeader.execution,ssz.capella.ExecutionPayloadHeader.defaultValue())

also the ssz type for above will depend upon the fork, so you will actually have to do:

`const sszType =config.getForkExecutionTypes(lcHeader.beacon.slot).ExecutionPayloadHeader`
and then check `sszType.equals(lcHeader.execution,sszType.defaultValue())`

nazarhussain replied:
Updated the `sszType` logic. For the `isForkExecution` seems correct to me. If that's not correct then we may need to update `isForkExecution`.

g11tech replied:
bellatrix which is execution fork might not have `execution` in LCHeader and should just return instead of throwing error, also could you chain this with above confition on L151 with `if else`

nazarhussain replied:
I don't get your point. If `bellatrix` is an execution fork why would it not have `execution` in the header?

g11tech replied:
execution is only bundled with LightClient header on/after capella

nazarhussain replied:
@g11tech Updated the code to `isForkWithdrawals`

g11tech commented on packages/prover/src/utils/execution.ts:
```diff
@@ -0,0 +1,108 @@
+  handler: ELRequestMethod;
+  proofProvider: ProofProvider;
+}): Promise<ELResponse | undefined> {
+  await proofProvider.waitToBeReady();
```
should we race here against a timeout?

nazarhussain replied:
Added an action item to the list, better to be done on multiple places in a separate PR.

g11tech commented on packages/prover/src/utils/execution.ts:
```diff
@@ -0,0 +1,108 @@
+  handler: ELRequestMethod,
+  args: [address: string, storageKeys: string[], block: number | string]
+): Promise<ELProof> {
+  // TODO: Find better way to generate random id
```
shouldn't id be increasing sequence? i am not exactly sure what jsonrpc spec says

nazarhussain replied:
`jsonrpc` standard is flexible about the ID. The incremental behavior may collide with internal provider requests.

nazarhussain committed:
Update the code as per feedback

nazarhussain committed:
Fix few types in the tests

nazarhussain committed:
Move the usage to isForkWithdrawls

nazarhussain committed:
Fix a unit test

g11tech reviewed (approved):
approving with the caveat that followup PRs would need to polish and address some of the unaddressed issues here to productionize the `prover`

But an amazing effort here to bring such an amazing functionality to lodestar codebase! :clap:

nazarhussain commented:
> approving with the caveat that followup PRs would need to polish and address some of the unaddressed issues here to productionize the `prover`
> 
> But an amazing effort here to bring such an amazing functionality to lodestar codebase! 👏

All the remaining points are added as action item to this issue https://github.com/ChainSafe/lodestar/issues/4706 so we will not forget it.

wemeetagain commented:
:tada: This PR is included in [v1.8.0](https://github.com/ChainSafe/lodestar/releases/tag/v1.8.0) :tada:
